algorithms/rainbow/model.py

```diff
@@ -0,0 +1,112 @@
+"""
+Model definition definition for DQN and Noisy layers
+"""
+import math
+import torch
+from torch import nn
+from torch.nn import functional as F
+
+
+class RainbowDQN(nn.Module):
+    """
+    From the paper "Rainbow: Combining Improvements in Deep Reinforcement Learning"
+    Source: https://arxiv.org/pdf/1710.02298.pdf
+
+    Main model used as Q function estimator for policy generation
+    """
+    def __init__(self, args, action_space):
+        super().__init__()
+        self.atoms = args.atoms
+        self.action_space = action_space.n
+        self.linear_in = self.get_linear_size(args)
+
+        self.conv1 = nn.Conv2d(args.history_length, 32, 8, stride=4, padding=1)
+        self.conv2 = nn.Conv2d(32, 64, 4, stride=2)
+        self.conv3 = nn.Conv2d(64, 64, 3)
+        self.fc_h_v = NoisyLinear(self.linear_in, args.hidden_size, std_init=args.noisy_std)
+        self.fc_h_a = NoisyLinear(self.linear_in, args.hidden_size, std_init=args.noisy_std)
+        self.fc_z_v = NoisyLinear(args.hidden_size, self.atoms, std_init=args.noisy_std)
+        self.fc_z_a = NoisyLinear(args.hidden_size, self.action_space * self.atoms, std_init=args.noisy_std)
+
+    def forward(self, x, log=False):
+        x = F.relu(self.conv1(x))
+        x = F.relu(self.conv2(x))
+        x = F.relu(self.conv3(x))
+        x = x.view(-1, self.linear_in)
+        v = self.fc_z_v(F.relu(self.fc_h_v(x)))  # Value stream
+        a = self.fc_z_a(F.relu(self.fc_h_a(x)))  # Advantage stream
+        v, a = v.view(-1, 1, self.atoms), a.view(-1, self.action_space, self.atoms)
+        q = v + a - a.mean(1, keepdim=True)  # Combine streams
+        if log:  # Use log softmax for numerical stability
+          q = F.log_softmax(q, dim=2)  # Log probabilities with action over second dimension
+        else:
+          q = F.softmax(q, dim=2)  # Probabilities with action over second dimension
```

Comment:
Ok I can walk you through the process with docstrings, but it might be overdocumenting. The naming convention here also doesn't help, because before the softmax we don't have "q" but "z". "z" includes atoms and this softmax is taken over each Z(s,a). Will change the name and document better